File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,bool>Map;
        int n=s.size(),st=0,ans=0;
        for(int i=0;i<n;i++){
            if(Map.find(s[i])!=Map.end()){
                while(st<i){
                    Map.erase(s[st]);
                    if(s[st]==s[i]){
                        st++;
                        break;
                    }
                    st++;
                }
                Map[s[i]]=true;
            }
            else{
                Map[s[i]]=true;
            }
            ans=max(ans,i-st+1);
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n=len(s)
        flag=defaultdict(bool)
        st=0
        ans=0
        for i in range(n):
            if s[i] in flag:
                while s[st]!=s[i]:
                    del flag[s[st]]
                    st+=1
                st+=1
            flag[s[i]]=True
            ans=max(ans,(i-st+1))   

        return ans       


        